File: Java/src/main/java/org/onlineoptimisation/problem/benchmark/AttractiveSectorFunction.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Mat;
import org.armadillojava.Op;

public class AttractiveSectorFunction extends BenchmarkProblem {

  protected Mat _Q;
  protected Mat _R;
  
  public AttractiveSectorFunction(int numberOfDimensions) {
    super(numberOfDimensions);
    
    _Q = randomRotation();
    _R = randomRotation();
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    Col z = _Q.times(delta(10).times(_R.times(parameter)));
    z.elem(Arma.find((z.elemTimes(_parameterShift)).strictGreaterThan(0)), Op.TIMES, 100);

    return oscillate(Math.pow(Arma.accu(Arma.square(z)), 0.9));
  }

}

File: src/optimisationProblem/benchmark/attractiveSectorFunction.cpp
#include <hop_bits/optimisationProblem/benchmark/attractiveSectorFunction.hpp>

namespace hop {
  double AttractiveSectorFunction::getObjectiveValueImplementation(const arma::Col<double>& parameter) const {
    arma::Col<double> z = rotationQ_ * (delta_ % (rotationR_ * (parameter - translation_)));
    z.elem(arma::find(z % translation_ > 0)) *= 100.0;

    return std::pow(getOscillationTransformation(std::pow(norm(z), 2)), 0.9);
  }
}
